### PR TITLE
[ParticleMechanics] New arrangement of strategy and particle generator

### PR DESCRIPTION
**Pull request by bodhinandach**
The current version of MPM strategy constructor requires a string of geometry type and an int of particle per element to be passed from the solver. This PR change and optimize it in such a way that the following can be achieved:
- For different submodelparts of the initial mesh, a different type of geometry can be used, e.g. one triangular and one quad.
- For each submodelpart of the initial mesh, we can assign different number of particle per element, e.g. at region where you want to have more particle, then it can be done simply by modifying from the json level.
- The geometry type of the initial mesh and the background can be different although the background can only have one type of mesh.
- User doesnt have to supply any type of element from the json, as it is now automatic.

I make it so that the old problem types can still work and put some warning already for the modification.
Further modification will be made in the interface following this PR.

### CODE REVIEW

**Review by antonialarese — COMMENTED** [review]
This seems fine to me. @philbucher do you see anything that need to be improved/changed?

**Comment by bodhinandach** [comment]
@philbucher @jcotela are there any comments from your sides?

File: applications/ParticleMechanicsApplication/custom_python/particle_mechanics_python_application.cpp [hunk 1]
```diff
@@ -119,6 +119,7 @@ namespace Python{
         KRATOS_REGISTER_IN_PYTHON_VARIABLE(m, MP_ACCUMULATED_PLASTIC_VOLUMETRIC_STRAIN );
         KRATOS_REGISTER_IN_PYTHON_VARIABLE(m, MP_ACCUMULATED_PLASTIC_DEVIATORIC_STRAIN );
         KRATOS_REGISTER_IN_PYTHON_VARIABLE(m, MP_MATERIAL_ID);
+        KRATOS_REGISTER_IN_PYTHON_VARIABLE(m, PARTICLE_PER_ELEMENT);
```
Comment (philbucher) [inline applications/ParticleMechanicsApplication/custom_python/particle_mechanics_python_application.cpp]: would `PARTICLES_PER_ELEMENT` make more sense?

File: applications/ParticleMechanicsApplication/custom_strategies/strategies/MPM_residual_based_newton_raphson_strategy.hpp [hunk 2]
```diff
@@ -573,7 +573,7 @@ class MPMResidualBasedNewtonRaphsonStrategy
             }
             else
             {
-                KRATOS_INFO("MPM_Strategy") << "WARNING: no free DOFs!! " << std::endl;
+                KRATOS_WARNING("MPM_Strategy") << "WARNING: no free DOFs!! " << std::endl;
```
Comment (philbucher) [inline applications/ParticleMechanicsApplication/custom_strategies/strategies/MPM_residual_based_newton_raphson_strategy.hpp:576]: I think this is an error in the core strategies ...?

File: applications/ParticleMechanicsApplication/custom_strategies/strategies/MPM_strategy.h [hunk 3]
```diff
@@ -217,11 +219,23 @@ class MPMStrategy
...
+                    }
+                    else{
+                        std::string warning_msg = "PARTICLE_PER_ELEMENT is not specified in Properties, ";
+                        warning_msg += "1 Particle per element is assumed.";
+                        KRATOS_WARNING("MPM_Strategy") << "WARNING: " << warning_msg << std::endl;
+                        particle_per_element = 1;
+                    }
+
                     Geometry< Node < 3 > >& rGeom = i->GetGeometry(); // current element's connectivity
+                    GeometryData::KratosGeometryType rGeoType = rGeom.GetGeometryType();
```
Comment (philbucher) [inline applications/ParticleMechanicsApplication/custom_strategies/strategies/MPM_strategy.h]: const

File: applications/ParticleMechanicsApplication/custom_strategies/strategies/MPM_strategy.h [hunk 4]
```diff
@@ -217,11 +219,23 @@ class MPMStrategy
...
+                        particle_per_element = i->GetProperties()[PARTICLE_PER_ELEMENT];
+                    }
+                    else{
+                        std::string warning_msg = "PARTICLE_PER_ELEMENT is not specified in Properties, ";
+                        warning_msg += "1 Particle per element is assumed.";
+                        KRATOS_WARNING("MPM_Strategy") << "WARNING: " << warning_msg << std::endl;
+                        particle_per_element = 1;
+                    }
+
                     Geometry< Node < 3 > >& rGeom = i->GetGeometry(); // current element's connectivity
```
Comment (philbucher) [inline applications/ParticleMechanicsApplication/custom_strategies/strategies/MPM_strategy.h]: const

File: applications/ParticleMechanicsApplication/custom_strategies/strategies/MPM_residual_based_newton_raphson_strategy.hpp (hunk 2 above)
Comment (bodhinandach) [reply applications/ParticleMechanicsApplication/custom_strategies/strategies/MPM_residual_based_newton_raphson_strategy.hpp:576]: I am not sure about this... I can't find it at the solving_strategy.h

Comment (philbucher) [reply applications/ParticleMechanicsApplication/custom_strategies/strategies/MPM_residual_based_newton_raphson_strategy.hpp:576]: ok then it is in the builderandsolver

**Review by philbucher — APPROVED** [review]

File: applications/ParticleMechanicsApplication/custom_strategies/strategies/MPM_residual_based_newton_raphson_strategy.hpp (hunk 2 above)
Comment (bodhinandach) [reply applications/ParticleMechanicsApplication/custom_strategies/strategies/MPM_residual_based_newton_raphson_strategy.hpp:576]: Yes, it is there :+1:

**Pull request merged**